feat(api): allow Device Authorization Grant using custom login UI (#9387)
zitadel/zitadel: docs/docs/guides/integrate/login-ui/device-auth.mdx

---
title: Support for the Device Authorization Grant in a Custom Login UI
sidebar_label: Device Authorization
---

In case one of your applications requires the OAuth2 Device Authorization Grant this guide will show you how to implement

> this in your application as well as the custom login UI.

Reviewer: Perhaps link to the device specific guide as well?

```suggestion
In case one of your applications requires the [OAuth2 Device Authorization Grant](/docs/guides/integrate/login/oidc/device-authorization) this guide will show you how to implement
```

Reviewer: Hmm this got connected to the wrong line. Gimme a sec to resend the suggestion.

The following flow shows you the different components you need to enable OAuth2 Device Authorization Grant for your login.
![Device Auth Flow](/img/guides/login-ui/device-auth-flow.png)

1. Your application makes a device authorization request to your login UI
2. The login UI proxies the request to ZITADEL.
3. ZITADEL parses the request and does what it needs to interpret certain parameters (e.g., organization scope, etc.)
4. ZITADEL returns the device authorization response

> 5. Your application presents the user_code and verification_uri or maybe even renders a QR code with the verification_uri_complete for the user to scan

Reviewer: ```suggestion
5. Your application presents the `user_code` and `verification_uri` or maybe even renders a QR code with the `verification_uri_complete` for the user to scan
```

9. Your login UI allows to approve or deny the device request
10. In case they approved, authenticate the user in your login UI by creating and updating a session with all the checks you need.
11. Inform ZITADEL about the decision:
    1. Authorize the device authorization request by sending the session and the previously retrieved id of the device authorization request to the ZITADEL API
    2. In case they denied, deny the device authorization from the ZITADEL API using the previously retrieved id of the device authorization request
12. Notify the user that they can close the window now and return to the application.
13. Your applications request to the token endpoint now receives the tokens or an error if the user denied the request.

## Example

Let's assume you host your login UI on the following URL:
```
https://login.example.com
```

## Device Authorization Request

A user opens your application and is unauthenticated, the application will create the following request:
```HTTP
POST /oauth/v2/device_authorization HTTP/1.1
Host: login.example.com
Content-type: application/x-www-form-urlencoded

> client_id=170086824411201793%40yourapp&

Reviewer: Nitpicking, but we no longer issue `<client_id>@<project>` form client IDs anymore.

```suggestion
client_id=170086824411201793&
```

scope=openid%20email%20profile
```

The request includes all the relevant information for the OAuth2 Device Authorization Grant and in this example we also have some scopes for the user.

You now have to proxy the auth request from your own UI to the device authorization Endpoint of ZITADEL.
For more information, see [OIDC Proxy](./typescript-repo#oidc-proxy) for the necessary headers.

:::note
The version and the optional custom URI for the available login UI is configurable under the application settings.
:::

The endpoint will return the device authorization response:
```json
{
  "device_code": "0jbAZbU3ClK-Mkt0li4U1A",
  "user_code": "FWRK-JGWK",
  "verification_uri": "https://login.example.com/device",
  "verification_uri_complete": "https://login.example.com/device?user_code=FWRK-JGWK",
  "expires_in": 300,
  "interval": 5
}
```

> Present the user_code and verification_uri or maybe even render a QR code with the verification_uri_complete for the user to scan.
> 
> Your login will have to provide a page on the verification_uri where the user can enter the user_code, or it's automatically filled in, if they scan the QR code.

Reviewer: ```suggestion
The device presents the `user_code` and `verification_uri` or maybe even render a QR code with the `verification_uri_complete` for the user to scan.

Your login will have to provide a page on the `verification_uri` where the user can enter the `user_code`, or it's automatically filled in, if they scan the QR code.
```

### Get the Device Authorization Request by User Code

With the user_code entered by the user you will now be able to get the information of the device authorization request.
[Get Device Authorization Request Documentation](/docs/apis/resources/oidc_service_v2/oidc-service-get-device-authorization-request)

```bash
curl --request GET \
  --url https://$ZITADEL_DOMAIN/v2/oidc/device_authorization/FWRK-JGWK \
  --header 'Authorization: Bearer '"$TOKEN"''
```

Response Example:

```json
{
  "deviceAuthorizationRequest": {
    "id": "uEwFVs5e5OOHrXuS/02WpwkN/2pucoIgHsFkktgMjDNl9gz/dt8",

>     "clientId": "170086824411201793@yourapp",

Author: ```suggestion
    "clientId": "170086824411201793",
```
